app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+            amount = balance
+        )
+        validate()
+    }
+
+    override fun onSubmit() {
+        // todo: https://p2pvalidator.atlassian.net/browse/PWN-9267
+    }
+
+    private fun setTokenAmount(targetTokenType: StrigaOffRampTokenType, amount: BigDecimal) {
+        when (targetTokenType) {
+            StrigaOffRampTokenType.TokenA -> {
+                inputAmountA = amount
+                val tokenAState = swapWidgetMapper.mapByState(
+                    tokenType = StrigaOffRampTokenType.TokenA,
+                    state = swapWidgetMapper.mapTokenA(
+                        amount = amount,
+                        balance = balance,
+                    )
+                )
+                view?.setTokenAWidgetState(tokenAState)
+            }
+            StrigaOffRampTokenType.TokenB -> {
+                inputAmountB = amount
+                val tokenBState = swapWidgetMapper.mapTokenB(amount)
+                view?.setTokenBWidgetState(
+                    swapWidgetMapper.mapByState(
+                        tokenType = StrigaOffRampTokenType.TokenB,
+                        state = tokenBState
+                    )
+                )
+            }
+        }
+    }
+
+    private fun setAndCalcTokenAmount(targetTokenType: StrigaOffRampTokenType, amount: BigDecimal) {
+        val calculatedAmount = interactor.calculateAmountByRate(targetTokenType, rate, amount)
+        setTokenAmount(targetTokenType, calculatedAmount)
+    }
+
+    private fun validate() {
+        val buttonState = interactor.validateAmount(
+            amountA = inputAmountA,
+            amountB = inputAmountB,
+            balance = balance
+        )
+        setButtonState(buttonState)
+
+        if (buttonState.isAmountError) {
+            view?.setTokenAErrorState(isError = true)
+        } else {
+            view?.setTokenAErrorState(isError = false)
+        }
+    }
+
+    private fun setButtonState(state: StrigaOffRampButtonState) {
+        timber.d("Set button state: $state")
+        viewState.value = viewState.value.copy(buttonState = state)
+    }
+
+    private fun setExchangeRateState(state: SwapRateTickerState) {
+        viewState.value = viewState.value.copy(exchangeRateState = state)
+    }
+
+    @NoCoverage
+    private fun logError(throwable: Throwable) {
+        when (throwable) {
+            is CancellationException -> {
+                if (!connectionManager.connectionStatus.value) {
+                    timber.i(throwable, "Canceled rate polling due to no internet connection")
+                } else {
+                    timber.i(
+                        throwable,
+                        "Canceled rate polling due to coroutine has been disposed"
+                    )
+                }
+            }
+            is StrigaDataLayerError.InternalError -> {
+                if (throwable.cause != null) {
+                    logError(throwable.cause!!)
```

Comment:
the reason why it is a recursion because the `cause` can be a throwable which we'd like to supporte in upper level, including `CancellationException` which is shouldn't be logged as error